src/parser/postgresparser.cpp

```diff
@@ -751,42 +749,45 @@ std::unique_ptr<SelectStatement> PostgresParser::SelectTransform(SelectStmt *roo
       }
       auto limit_desc = std::make_unique<LimitDescription>(limit, offset);
 
-      result = std::make_unique<SelectStatement>(std::move(target), select_distinct, std::move(from), std::move(where),
+      result = std::make_unique<SelectStatement>(std::move(target), select_distinct, std::move(from), where,
                                                  std::move(groupby), std::move(orderby), std::move(limit_desc));
       break;
     }
     case SETOP_UNION: {
-      result = SelectTransform(root->larg_);
-      result->SetUnionSelect(SelectTransform(root->rarg_));
+      result = SelectTransform(parse_result, root->larg_);
+      result->SetUnionSelect(SelectTransform(parse_result, root->rarg_));
       break;
     }
     default: {
-      // TODO(Wan): is Set the right message, or should it be Select?
-      PARSER_LOG_AND_THROW("SelectTransform", "Set operation", root->type_);
+      PARSER_LOG_AND_THROW("SelectTransform", "Select operation", root->type_);
     }
   }
 
   return result;
 }
 
 // Postgres.SelectStmt.targetList -> terrier.SelectStatement.select_
-std::vector<std::shared_ptr<AbstractExpression>> PostgresParser::TargetTransform(List *root) {
+std::vector<common::ManagedPointer<AbstractExpression>> PostgresParser::TargetTransform(ParseResult *parse_result,
+                                                                                        List *root) {
   // Postgres parses 'SELECT;' to nullptr
   if (root == nullptr) {
     throw PARSER_EXCEPTION("TargetTransform: root==null.");
   }
 
-  std::vector<std::shared_ptr<AbstractExpression>> result;
+  std::vector<common::ManagedPointer<AbstractExpression>> result;
   for (auto cell = root->head; cell != nullptr; cell = cell->next) {
     auto target = reinterpret_cast<ResTarget *>(cell->data.ptr_value);
-    result.emplace_back(ExprTransform(target->val_, target->name_));
+    auto expr = ExprTransform(parse_result, target->val_, target->name_);
+    auto expr_managed = common::ManagedPointer(expr);
+    parse_result->AddExpression(std::move(expr));
```

Comment:
This seems kind of awkward to me. Can there be situations where parse_result is deallocated before the result from this function is used. In those cases you would have segfaults

Reply:
Currently, it is the responsibility of the caller of BuildParseTree to keep the parsed results alive while people need to use it. How else would you manage this?